map/traffic_manager.hpp

```diff
@@ -40,46 +51,70 @@ class TrafficManager final
     {}
   };
 
+  using TrafficStateChangedFn = function<void(TrafficState)>;
   using GetMwmsByRectFn = function<vector<MwmSet::MwmId>(m2::RectD const &)>;
 
   TrafficManager(GetMwmsByRectFn const & getMwmsByRectFn, size_t maxCacheSizeBytes);
   ~TrafficManager();
 
+  void SetStateListener(TrafficStateChangedFn && onStateChangedFn);
+
   void SetEnabled(bool enabled);
 
   void UpdateViewport(ScreenBase const & screen);
   void UpdateMyPosition(MyPosition const & myPosition);
 
-  void OnRecover();
+  void OnDestroyGLContext();
+  void OnRecoverGLContext();
 
   void SetDrapeEngine(ref_ptr<df::DrapeEngine> engine);
 
 private:
   void ThreadRoutine();
   bool WaitForRequest(vector<MwmSet::MwmId> & mwms);
+
   void RequestTrafficData();
   void RequestTrafficData(MwmSet::MwmId const & mwmId);
+
   void OnTrafficDataResponse(traffic::TrafficInfo const & info);
+  void OnTrafficRequestFailed(MwmSet::MwmId const & mwmId);
+
+  void Clear();
   void CheckCacheSize();
 
-  bool m_isEnabled;
+  void UpdateState();
+  void ChangeState(TrafficState newState);
+  void NotifyStateChanged();
+
+  bool IsInvalidState() const;
+  bool IsEnabled() const;
 
   GetMwmsByRectFn m_getMwmsByRectFn;
 
   ref_ptr<df::DrapeEngine> m_drapeEngine;
 
-  MyPosition m_currentPosition;
-  ScreenBase m_currentModelView;
+  pair<MyPosition, bool> m_currentPosition = {MyPosition(), false};
+  pair<ScreenBase, bool> m_currentModelView = {ScreenBase(), false};
+
+  atomic<TrafficState> m_state;
+  atomic<bool> m_notifyStateChanged;
+  TrafficStateChangedFn m_onStateChangedFn;
```

Comment:
If this callback is always set from the UI thread, I suggest you to post a task on the UI thread which will call the callback.